### PR TITLE
chore: tune Fluent Bit logging performance [DET-4714]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change makes some tweaks to the fluent config to make it a little more reliable in the face of large log bursts and master or elastic backpressure.

## Test Plan
- [x] run with heavy logging experiments, monitor memory, cpu usage and watch fluent health 

## Commentary (optional)
(Old comment, retro below) This probably will matter more when we perf test elastic, but these are the things we're going to need to be tuning, probably. We probably should've been using `non-blocking` mode for the log driver all along, too, but using `non-blocking` without turning on the mmap buffers to the filesystem escape hatch means where we might've crashed the job before we would just lose logs --- with the escape hatch we should be able to recover unless we get too many logs for too long.

As remarked in https://determinedai.atlassian.net/browse/DET-4714, not running in non-blocking mode lead to very strange and unavoidable crashes within fluentbit when trying to grab logs from the daemon. To fix this, this PR moves to non-blocking mode and tweaks some others stuff that increased throughput in benchmarking.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: agent/internal/fluent.go [hunk 1]
```diff
@@ -292,6 +305,12 @@ func startLoggingContainer(
 				Memory:   1 << 30,
 				NanoCPUs: 1000000000,
 			},
+			LogConfig: container.LogConfig{
+				Type: "json-file",
+				Config: map[string]string{
+					"max-size": "10m",
```
Comment (stoksc) [inline agent/internal/fluent.go:311]: this is unrelated, just so the fluentbit container itself will do log rotation since just it's `INFO` flush logs log about a 1kb/s since we're flushing every .05 seconds.. it would take like 2 months to be 2gb of log space but still 🤷

File: master/internal/api/logs.go [hunk 2]
```diff
@@ -130,6 +125,10 @@ func (l *LogStoreProcessor) Receive(ctx *actor.Context) error {
 			}
 			actors.NotifyAfter(ctx, 500*time.Millisecond, tick{})
 		default:
+			if l.ctx.Err() != nil {
```
Comment (stoksc) [inline master/internal/api/logs.go:128]: Saw a situation occasionally where the ctx was checked so far from the send where we sent on a dead gRPC channel, so I moved it up, to where it matters.

File: tools/docker-compose.yaml [hunk 3]
```diff
@@ -21,3 +21,4 @@ services:
     environment:
       discovery.type: 'single-node'
       cluster.routing.allocation.disk.threshold_enabled: 'false'
+      logger.level: 'WARN'
```
Comment (stoksc) [inline tools/docker-compose.yaml:24]: Those logs were getting really annoying, this catches anything that matters.

File: agent/internal/fluent.go [hunk 4]
```diff
@@ -130,6 +130,15 @@ func startLoggingContainer(
 		[]fluent.ConfigSection{
 			{
 				{"Name", "forward"},
+				// mem_buf_limit in combination with storage.type filesystem allows
+				// fluentbit to store full buffers in the filesystem if the rest of the
+				// pipeline is backed up. These both in combination with allowing the
+				// docker log driver to run in non-blocking mode allows us to ship
+				// without impacting application performance in bursts.
+				// This scheme is described in more detail at:
+				// https://docs.fluentbit.io/manual/administration/buffering-and-storage
```
Comment (dzhu) [inline agent/internal/fluent.go:139]:
```suggestion
				// Setting mem_buf_limit and storage.type=filesystem allows Fluent Bit to buffer log data to
				// disk if the rest of the pipeline is backed up. In combination with setting the Docker log
				// driver to run in non-blocking mode, that lets us avoid impacting application performance when
				// there are bursts in log output.
				//
				// This scheme is described in more detail at:
				// https://docs.fluentbit.io/manual/administration/buffering-and-storage
```

**Review by dzhu — COMMENTED** [review]
Thanks for investigating all of this stuff!

File: master/internal/api/logs.go (hunk 2 above)
Comment (stoksc) [reply master/internal/api/logs.go:128]: This `ctx` from the gRPC request. If the stream has closed, `ctx.Err()` is non nil and we stop. We were previously checking it before we fetched the logs. Where it was we were really likely to check, fetch logs and send on a dead channel then check at the right time (since most of the time is spent in fetching, a change usually happens while we're doing that). Though now that I've typed it out, I realize where it is we _always_ do the fetch, even if we're canceled, and then just throw it away instead of erroring, which is wasteful. So I'm going to put it both places, one to avoid wasting work, another to avoid bloating the logs with "failed to send on yada yada.."

**Review by dzhu — APPROVED** [review]
